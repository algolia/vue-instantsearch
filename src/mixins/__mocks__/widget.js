let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
