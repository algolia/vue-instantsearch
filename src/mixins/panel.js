import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
