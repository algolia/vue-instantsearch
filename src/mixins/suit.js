import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
