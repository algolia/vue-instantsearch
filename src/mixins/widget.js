const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
