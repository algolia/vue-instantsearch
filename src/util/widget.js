export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
