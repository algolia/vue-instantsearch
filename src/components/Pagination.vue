<template>
  <div
    v-if="state"
    :class="suit()"
  >
    <slot
      :refine="refine"
      :createURL="state.createURL"
      :currentRefinement="state.currentRefinement"
      :nbHits="state.nbHits"
      :nbPages="state.nbPages"
      :pages="state.pages"
      :isFirstPage="state.isFirstPage"
      :isLastPage="state.isLastPage"
    >
      <ul :class="suit('list')">
        <li
          :class="{
            [suit('item')]:true,
            [suit('item', 'firstPage')]: true,
            [suit('item', 'disabled')]: state.isFirstPage
          }"
        >
          <template v-if="!state.isFirstPage">
            <a
              :class="suit('link')"
              aria-label="First"
              :href="state.createURL(0)"
              @click.prevent="refine(0)"
            >‹‹</a>
          </template>
          <template v-else>
            <span
              :class="suit('link')"
              aria-label="Previous"
            >
              ‹‹
            </span>
          </template>
        </li>
        <li
          :class="{
            [suit('item')]: true,
            [suit('item', 'previousPage')]: true,
            [suit('item', 'disabled')]: state.isFirstPage
          }"
        >
          <template v-if="!state.isFirstPage">
            <a
              :class="suit('link')"
              aria-label="Previous"
              :href="state.createURL(state.currentRefinement - 1)"
              @click.prevent="refine(state.currentRefinement - 1)"
            >
              ‹
            </a>
          </template>
          <template v-else>
            <span
              :class="suit('link')"
              aria-label="Previous"
            >
              ‹
            </span>
          </template>
        </li>

        <li
          :class="{
            [suit('item')]:true,
            [suit('item', 'selected')]: state.currentRefinement === page
          }"
          v-for="page in state.pages"
          :key="page"
        >
          <a
            :class="suit('link')"
            :href="state.createURL(page)"
            @click.prevent="refine(page)"
          >
            {{ page + 1 }}
          </a>
        </li>

        <li
          :class="{
            [suit('item')]:true,
            [suit('item','nextPage')]: true,
            [suit('item','disabled')]: state.isLastPage
          }"
        >
          <template v-if="!state.isLastPage">
            <a
              :class="suit('link')"
              aria-label="Next"
              :href="state.createURL(state.currentRefinement + 1)"
              @click.prevent="refine(state.currentRefinement + 1)"
            >›</a>
          </template>
          <template v-else>
            <span
              :class="suit('link')"
              aria-label="Next"
            >
              ›
            </span>
          </template>
        </li>
        <li
          :class="{
            [suit('item')]:true,
            [suit('item','lastPage')]: true,
            [suit('item','disabled')]: state.isLastPage
          }"
        >
          <template v-if="!state.isLastPage">
            <a
              :class="suit('link')"
              aria-label="Last"
              :href="state.createURL(state.nbPages - 1)"
              @click.prevent="refine(state.nbPages - 1)"
            >››</a>
          </template>
          <template v-else>
            <span
              :class="suit('link')"
              aria-label="Last"
            >
              ››
            </span>
          </template>
        </li>
      </ul>
    </slot>
  </div>
</template>

<script>
import { connectPagination } from 'instantsearch.js/es/connectors';
import { createPanelConsumerMixin } from '../mixins/panel';
import { createWidgetMixin } from '../mixins/widget';

export default {
  mixins: [
    createWidgetMixin({ connector: connectPagination }),
    createPanelConsumerMixin({
      mapStateToCanRefine: state => state.nbPages > 1,
    }),
  ],
  props: {
    padding: {
      type: Number,
      default: 3,
      validator(value) {
        return value > 0;
      },
    },
    totalPages: {
      type: Number,
      default: undefined,
      validator(value) {
        return value > 0;
      },
    },
    showFirst: {
      type: Boolean,
      default: true,
    },
    showLast: {
      type: Boolean,
      default: true,
    },
    showNext: {
      type: Boolean,
      default: true,
    },
    showPrevious: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      widgetName: 'Pagination',
    };
  },
  computed: {
    widgetParams() {
      return {
        padding: this.padding,
        maxPages: this.totalPages,
      };
    },
  },
  methods: {
    refine(page) {
      const p = Math.min(Math.max(page, 0), this.state.nbPages - 1);
      this.state.refine(p);
      // @TODO: do this in a general way
      this.$emit('page-change', p);
    },
  },
};
</script>
