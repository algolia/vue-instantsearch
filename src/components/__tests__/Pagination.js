import { mount } from '@vue/test-utils';
import { __setState } from '../../mixins/widget';
import Pagination from '../Pagination.vue';

jest.mock('../../mixins/widget');
jest.mock('../../mixins/panel');

const defaultState = {
  createURL: () => '#',
  refine: jest.fn(),
  pages: [0, 1, 2, 3, 4, 5, 6],
  nbPages: 20,
  currentRefinement: 0,
  isFirstPage: true,
  isLastPage: false,
};

it('accepts a padding prop', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(Pagination, {
    propsData: {
      padding: 5,
    },
  });

  expect(wrapper.vm.widgetParams.padding).toBe(5);
});

it('accepts a totalPages prop', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(Pagination, {
    propsData: {
      totalPages: 10,
    },
  });

  expect(wrapper.vm.widgetParams.totalPages).toBe(10);
});

it('renders correctly first page', () => {
  __setState(defaultState);
  const wrapper = mount(Pagination);

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly another page', () => {
  __setState({
    ...defaultState,
    pages: [3, 4, 5, 6, 7, 8, 9],
    nbPages: 20,
    currentRefinement: 6,
    isFirstPage: false,
    isLastPage: false,
  });
  const wrapper = mount(Pagination);

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly last page', () => {
  __setState({
    ...defaultState,
    pages: [3, 4, 5, 6, 7, 8, 9],
    nbPages: 9,
    currentRefinement: 9,
    isFirstPage: false,
    isLastPage: true,
  });
  const wrapper = mount(Pagination);

  expect(wrapper.html()).toMatchSnapshot();
});

it('Moves to the first page on that button', () => {
  __setState({
    ...defaultState,
    isFirstPage: false,
  });
  const wrapper = mount(Pagination);

  const firstPage = wrapper.find(
    '.ais-Pagination-item--firstPage .ais-Pagination-link'
  );
  firstPage.trigger('click');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith(0);
});

it('Moves to the next page on that button', () => {
  const currentRefinement = 5;
  __setState({
    ...defaultState,
    currentRefinement,
  });
  const wrapper = mount(Pagination);

  const nextPage = wrapper.find(
    '.ais-Pagination-item--nextPage .ais-Pagination-link'
  );
  nextPage.trigger('click');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith(
    currentRefinement + 1
  );
});

it('Moves to the last page on that button', () => {
  const nbPages = 1000;
  __setState({
    ...defaultState,
    isLastPage: false,
    nbPages,
  });
  const wrapper = mount(Pagination);

  const lastPage = wrapper.find(
    '.ais-Pagination-item--lastPage .ais-Pagination-link'
  );
  lastPage.trigger('click');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith(nbPages - 1);
});

it('Moves to the previous page on that button', () => {
  const currentRefinement = 5;
  __setState({
    ...defaultState,
    isFirstPage: false,
    currentRefinement,
  });
  const wrapper = mount(Pagination);

  const previousPage = wrapper.find(
    '.ais-Pagination-item--previousPage .ais-Pagination-link'
  );
  previousPage.trigger('click');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith(
    currentRefinement - 1
  );
});

it('calls the Panel mixin with `nbPages`', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Pagination);

  const mapStateToCanRefine = () =>
    wrapper.vm.mapStateToCanRefine(wrapper.vm.state);

  expect(mapStateToCanRefine()).toBe(true);

  wrapper.setData({
    state: {
      nbPages: 1,
    },
  });

  expect(mapStateToCanRefine()).toBe(false);
});

it('implements showFirst', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Pagination, {
    propsData: {
      showFirst: false,
    },
  });

  expect(wrapper.find('.ais-Pagination-item--firstPage').exists()).toBe(false);

  wrapper.setProps({
    showFirst: true,
  });

  expect(wrapper.find('.ais-Pagination-item--firstPage').exists()).toBe(true);
});

it('implements showPrevious', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Pagination, {
    propsData: {
      showPrevious: false,
    },
  });

  expect(wrapper.find('.ais-Pagination-item--previousPage').exists()).toBe(
    false
  );

  wrapper.setProps({
    showPrevious: true,
  });

  expect(wrapper.find('.ais-Pagination-item--previousPage').exists()).toBe(
    true
  );
});

it('implements showLast', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Pagination, {
    propsData: {
      showLast: false,
    },
  });

  expect(wrapper.find('.ais-Pagination-item--lastPage').exists()).toBe(false);

  wrapper.setProps({
    showLast: true,
  });

  expect(wrapper.find('.ais-Pagination-item--lastPage').exists()).toBe(true);
});

it('implements showNext', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Pagination, {
    propsData: {
      showNext: false,
    },
  });

  expect(wrapper.find('.ais-Pagination-item--nextPage').exists()).toBe(false);

  wrapper.setProps({
    showNext: true,
  });

  expect(wrapper.find('.ais-Pagination-item--nextPage').exists()).toBe(true);
});
