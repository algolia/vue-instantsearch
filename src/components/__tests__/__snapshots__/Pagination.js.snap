// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly another page 1`] = `

<div class="ais-Pagination">
  <ul class="ais-Pagination-list">
    <li class="ais-Pagination-item ais-Pagination-item--firstPage">
      <a aria-label="First"
         href="#"
         class="ais-Pagination-link"
      >
        ‹‹
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--previousPage">
      <a aria-label="Previous"
         href="#"
         class="ais-Pagination-link"
      >
        ‹
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        4
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        5
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        6
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--selected">
      <a href="#"
         class="ais-Pagination-link"
      >
        7
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        8
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        9
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        10
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--nextPage">
      <a aria-label="Next"
         href="#"
         class="ais-Pagination-link"
      >
        ›
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--lastPage">
      <a aria-label="Last"
         href="#"
         class="ais-Pagination-link"
      >
        ››
      </a>
    </li>
  </ul>
</div>

`;

exports[`renders correctly first page 1`] = `

<div class="ais-Pagination">
  <ul class="ais-Pagination-list">
    <li class="ais-Pagination-item ais-Pagination-item--firstPage ais-Pagination-item--disabled">
      <span aria-label="First"
            class="ais-Pagination-link"
      >
        ‹‹
      </span>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--previousPage ais-Pagination-item--disabled">
      <span aria-label="Previous"
            class="ais-Pagination-link"
      >
        ‹
      </span>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--selected">
      <a href="#"
         class="ais-Pagination-link"
      >
        1
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        2
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        3
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        4
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        5
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        6
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        7
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--nextPage">
      <a aria-label="Next"
         href="#"
         class="ais-Pagination-link"
      >
        ›
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--lastPage">
      <a aria-label="Last"
         href="#"
         class="ais-Pagination-link"
      >
        ››
      </a>
    </li>
  </ul>
</div>

`;

exports[`renders correctly last page 1`] = `

<div class="ais-Pagination">
  <ul class="ais-Pagination-list">
    <li class="ais-Pagination-item ais-Pagination-item--firstPage">
      <a aria-label="First"
         href="#"
         class="ais-Pagination-link"
      >
        ‹‹
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--previousPage">
      <a aria-label="Previous"
         href="#"
         class="ais-Pagination-link"
      >
        ‹
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        4
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        5
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        6
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        7
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        8
      </a>
    </li>
    <li class="ais-Pagination-item">
      <a href="#"
         class="ais-Pagination-link"
      >
        9
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--selected">
      <a href="#"
         class="ais-Pagination-link"
      >
        10
      </a>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--nextPage ais-Pagination-item--disabled">
      <span aria-label="Next"
            class="ais-Pagination-link"
      >
        ›
      </span>
    </li>
    <li class="ais-Pagination-item ais-Pagination-item--lastPage ais-Pagination-item--disabled">
      <span aria-label="Last"
            class="ais-Pagination-link"
      >
        ››
      </span>
    </li>
  </ul>
</div>

`;
